gtfs_rt_validator_api.py

```diff
@@ -25,42 +72,169 @@ def validate(gtfs_file, rt_path, out_file=None, verbose=False):
+        out_dir: a directory to store fetched files and results in.
+        results_dir: a bucket path to copy results to.
+        verbose: whether to print helpful messages along the way.
+
+    Note that if out_dir is unspecified, the validation occurs in a temporary directory.
+        
+    """
 
     import gcsfs
-    import shutil
 
     fs = gcsfs.GCSFileSystem(project_id, token=token)
 
-    with TemporaryDirectory() as tmp_dir:
-        path_gtfs = f"{tmp_dir}/gtfs"
-        path_gtfs_zip = f"{path_gtfs}.zip"
-        path_rt = f"{tmp_dir}/rt"
+    if not out_dir:
+        tmp_dir = TemporaryDirectory()
+        tmp_dir_name = tmp_dir.name
+    else:
+        tmp_dir = None
+        tmp_dir_name = out_dir
+
+    if results_bucket and not results_bucket.endswith("/"):
+        results_bucket = f"{results_bucket}/"
+
+    final_json_dir = Path(tmp_dir_name) / "newline_json"
+
+    try:
+        print("Fetching data")
+
+        dst_path_gtfs = f"{tmp_dir_name}/gtfs"
+        dst_path_rt = f"{tmp_dir_name}/rt"
 
         # fetch and zip gtfs schedule
-        fs.get(gtfs_schedule_path, path_gtfs, recursive=True)
-        shutil.make_archive(path_gtfs, "zip", path_gtfs)
+        download_gtfs_schedule_zip(gtfs_schedule_path, dst_path_gtfs, fs)
 
         # fetch rt data
-        # TODO: replace with gtfs_rt_glob
-        if gtfs_rt_date:
-            fs.glob(f"{gtfs_rt_date}T*/")
+        # TODO: remove hard-coding
+        if gtfs_rt_glob_path is None:
+            raise ValueError("One of gtfs rt glob path or date must be specified")
+
+
+        download_rt_files(dst_path_rt, fs, glob_path = gtfs_rt_glob_path)
+
+        print("Validating data")
+        validate(f"{dst_path_gtfs}.zip", dst_path_rt, verbose=verbose)
+
+        if results_bucket:
+            # validator stores results as {filename}.results.json
+            print(f"Putting data into results bucket: {results_bucket}")
+
+            # fetch all results files created by the validator
+            all_results = list(Path(dst_path_rt).glob("*.results.json"))
+
+            
+            final_json_dir.mkdir(exist_ok=True)
+            final_files = []
+            for result in all_results:
+                # we appended a final timestamp to the files so that the validator
+                # can use it to order them during validation. here, we remove that
+                # timestamp, so we can use a single wildcard to select, eg..
+                # *trip_updates.results.json
+                result_out = "__".join(result.name.split("__")[:-1])
+
+                json_to_newline_delimited(result, final_json_dir / result_out)
+                final_files.append(final_json_dir / result_out)
+            
+
+            fs.put(final_files, results_bucket)
+
+    except Exception as e:
+        if isinstance(tmp_dir, TemporaryDirectory):
+            tmp_dir.cleanup()
+
+        raise e
```

Comment:
Is the temporary directory cleanup still needed even if an exception occurs? If so, that cleanup should happen in a finally clause:

```suggestion
    except Exception as e:
        raise e
    finally:
        if isinstance(tmp_dir, TemporaryDirectory):
            tmp_dir.cleanup()
```

Reply:
Ah, thanks